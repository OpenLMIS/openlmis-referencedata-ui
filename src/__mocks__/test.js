import React from 'react';

export const Component = () => {
    return <div>Mocked Component</div>;
};
