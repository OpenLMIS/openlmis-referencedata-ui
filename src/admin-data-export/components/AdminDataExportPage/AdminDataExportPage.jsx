/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2017 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms
 * of the GNU Affero General Public License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
 * See the GNU Affero General Public License for more details. You should have received a copy of
 * the GNU Affero General Public License along with this program. If not, see
 * http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

import React, { useEffect, useState, useMemo } from 'react';
import { toast } from 'react-toastify';

import getService from '../../../react-components/utils/angular-utils';
import Select from '../../../react-components/inputs/select';
import MultiSelect from '../../../react-components/inputs/multi-select';
import { DATA_EXPORT, TYPE_OF_EXPORTS } from '../../consts';
import Loading from '../../../react-components/modals/loading';

const AdminDataExportPage = () => {

    const menu = document.getElementsByClassName('header ng-scope')[0];
    const [typeOfExport, setTypeOfExport] = useState("");
    const [optionsForDatas, setOptionsForDatas] = useState(DATA_EXPORT);
    const [selectedFiles, setSelectedFiles] = useState([]);
    const [displayLoading, setDisplayLoading] = useState(false);

    const serverService = useMemo(
        () => {
            return getService('adminDataExport');
        },
        []
    );

    const toggleOption = ({ id }) => {
      setSelectedFiles(prevSelected => {
        const newArray = [...prevSelected]
        if (newArray.includes(id)) {
            return newArray.filter(item => item != id)
        } else {
            newArray.push(id);
            return newArray;
        }
      })
    }

    const downloadZip = () => {
      if (selectedFiles.length > 0) {

        const adjustedElements = selectedFiles.map(element => element.replace('.csv', ''));

        const data = adjustedElements.toString();
        const token = localStorage.getItem('openlmis.ACCESS_TOKEN');

        setDisplayLoading(true);
        serverService.exportData(data)
          .then(() => {
            window.open(
                `${serverService.urlFactory(`/api/exportData?format=csv&data=${data}&access_token=${token}`)}`,
                "_blank"
            );
            setDisplayLoading(false);
            toast.success('Data has been exported correctly');
        });

        setTypeOfExport("");
        setSelectedFiles([]);
      }
    };

    useEffect(() => menu.style.display = '', [menu]);

    useEffect(() => {
      if (typeOfExport) {
        setOptionsForDatas(prevOptionsForFiles => {
          const newOptionsForFiles = [...DATA_EXPORT];
          return newOptionsForFiles.filter(
            data => data.type === typeOfExport
          );
        });
      } else {
        setTypeOfExport("");
        setOptionsForDatas(DATA_EXPORT);
      }
    }, [typeOfExport]);

    return (
        <>
            <div>
                <h2 id='data-export-header'>
                  Data Export
                </h2>
                <div className='required' style={{marginBottom: '4px', fontFamily: 'Arial', fontSize: '16px'}}>
                  <label id='type-header'>
                    Type
                  </label>
                </div>
                <div className='field-full-width' style={{marginBottom: '16px', width: '40%'}}>
                    <Select
                      options={TYPE_OF_EXPORTS}
                      onChange={value => {
                        setSelectedFiles([]);
                        setTypeOfExport(value);
                      }}
                      value={typeOfExport}
                    />
                </div>

                <div className='required' style={{marginBottom: '4px', fontFamily: 'Arial', fontSize: '16px'}}>
                  <label id='type-header'>
                    Data
                  </label>
                </div>
                <div className='field-full-width' style={{marginBottom: '16px', width: '40%'}}>
                    <MultiSelect
                      options={optionsForDatas}
                      selected={selectedFiles}
                      toggleOption={toggleOption}
                      disabled={!typeOfExport}
                    />
                </div>
            </div>
            <div className="openlmis-toolbar">
                <button
                  className='primary'
                  type='button'
                  style={{ marginTop: '0.5em' }}
                  disabled={selectedFiles.length === 0}
                  onClick={downloadZip}
                >
                  Export
                </button>
            </div>
            <Loading display={displayLoading}/>
        </>
    )
};

export default AdminDataExportPage;
