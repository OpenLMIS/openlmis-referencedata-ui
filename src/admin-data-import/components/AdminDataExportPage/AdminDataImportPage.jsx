/*
 * This program is part of the OpenLMIS logistics management information system platform software.
 * Copyright © 2017 VillageReach
 *
 * This program is free software: you can redistribute it and/or modify it under the terms
 * of the GNU Affero General Public License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. 
 * See the GNU Affero General Public License for more details. You should have received a copy of
 * the GNU Affero General Public License along with this program. If not, see
 * http://www.gnu.org/licenses.  For additional information contact info@OpenLMIS.org. 
 */

import React, { useState, useMemo } from 'react';
import { toast } from 'react-toastify';
import getService from '../../../react-components/utils/angular-utils';
import Select from '../../../react-components/inputs/select';
import Loading from '../../../react-components/modals/loading';
import {TYPE_OF_IMPORTS} from '../../consts';

const AdminDataImportPage = () => {

    const [typeOfImport, setTypeOfImport] = useState('');
    const [selectedFile, setSelectedFile] = useState('');
    const [displayLoading, setDisplayLoading] = useState(false);

    const serverService = useMemo(
        () => {
            return getService('adminDataImport');
        },
        []
    );

    const importZip = () => {
        if (selectedFile) {
            setDisplayLoading(true);
            serverService.importData(selectedFile)
                .then(() => toast.success('Data has been imported correctly'))
                .finally(() => {
                    setSelectedFile('');
                    setTypeOfImport('');
                    setDisplayLoading(false);
                });
        }
    }

    const FileUploader = () => {
        const hiddenFileInput = React.useRef(null);

        const handleClick = () => {
            hiddenFileInput.current.click();
        };

        const handleChange = event => {
            const fileUploaded = event.target.files[0];
            setSelectedFile(fileUploaded);
        };

        return (
            <>
                {!selectedFile &&
                    <>
                        <button
                            className='file-upload-button'
                            onClick={handleClick}
                            disabled={!typeOfImport}
                        >
                            Select File
                        </button>
                        <input
                        style={{display: 'none'}}
                        ref={hiddenFileInput}
                        type='file'
                        onChange={handleChange}
                        disabled={!typeOfImport}
                        accept=".zip, .rar, .7z"
                        />
                    </>
                }
                {selectedFile &&
                    <div className='input-import-file'>
                        <input
                            type='text'
                            value={selectedFile.name}
                        />
                        <button type='button'>
                            <i className="fa-duotone fa-xmark"></i>
                            <i
                                className='fa fa-times clear-icon clear clear-button'
                                onClick={() => setSelectedFile('')}
                            />
                        </button>
                    </div>
                }
            </>
        );
    }

    return (
        <>
            <div>
                <h2 id='data-export-header'>
                  Data Import
                </h2>
                <div className='required' style={{marginBottom: '4px', fontFamily: 'Arial', fontSize: '16px'}}>
                  <label id='type-header'>
                    Type
                  </label>
                </div>
                <div className='field-full-width' style={{marginBottom: '16px', width: '40%'}}>
                    <Select
                      options={TYPE_OF_IMPORTS}
                      onChange={value => {
                        setSelectedFile('');
                        setTypeOfImport(value);
                      }}
                      value={typeOfImport}
                    />
                    {TYPE_OF_IMPORTS.map(type => {
                        if(type.value === typeOfImport){
                            return <p style={{maxWidth: '544px'}} key={type.value}>{type.info}</p>
                        }
                    })}
                </div>
                <div className='required' style={{marginBottom: '4px', fontFamily: 'Arial', fontSize: '16px'}}>
                  <label id='type-header'>
                   Zip File
                  </label>
                </div>
                <div className='field-full-width' style={{marginBottom: '16px', width: '40%'}}>
                    <FileUploader />
                </div>
            </div>
            <div className="openlmis-toolbar">
                <button
                  className='primary'
                  type='button'
                  style={{ marginTop: '0.5em' }}
                  disabled={selectedFile.length === 0}
                  onClick={importZip}
                >
                  Import
                </button>
            </div>
            {displayLoading &&
              <Loading/>
            }
        </>
    )
};

export default AdminDataImportPage;
