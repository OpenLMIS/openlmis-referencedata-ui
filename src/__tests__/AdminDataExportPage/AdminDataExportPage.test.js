import React from 'react';
import {fireEvent, render, screen} from '@testing-library/react';

import { Component } from "../../__mocks__/test.js";


import AdminDataExportPage from '../../admin-data-export/components/AdminDataExportPage/AdminDataExportPage.jsx'

jest.mock('../../react-components/utils/angular-utils', () => 'Angular Utils Mock')

describe('MyComponent', () => {
    it('renders correctly', () => {
        const { container, getByText } = render(<Component />);
        expect(getByText('Mocked Component')).toBeTruthy();


        expect(container).toMatchSnapshot()
    });
});

describe('Unit tests for the AdminDataExportPage.', () => {

    describe('Should match snapshot -', () => {
        test('create.', async () => {
            const { container, getByTestId } = render(
                <AdminDataExportPage />
            )

            expect(container).toMatchSnapshot()
        })
    })

})
