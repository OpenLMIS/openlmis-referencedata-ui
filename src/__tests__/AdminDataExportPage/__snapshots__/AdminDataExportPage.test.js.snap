// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyComponent renders correctly 1`] = `
<div>
  <div>
    Mocked Component
  </div>
</div>
`;

exports[`Unit tests for the AdminDataExportPage. Should match snapshot - create. 1`] = `<div />`;
